crates/cli/src/format.rs

```diff
@@ -256,10 +265,158 @@ fn fmt_all<'a>(buf: &mut Buf<'a>, ast: &'a FullAst) {
     buf.fmt_end_of_file();
 }
 
+pub fn annotate_file(arena: &Bump, file: PathBuf) -> Result<(), LoadingProblem> {
+    let load_config = LoadConfig {
+        target: Target::default(),
+        function_kind: FunctionKind::from_env(),
+        render: RenderTarget::ColorTerminal,
+        palette: DEFAULT_PALETTE,
+        threading: Threading::AllAvailable,
+        exec_mode: ExecutionMode::Check,
+    };
+
+    let mut loaded = roc_load::load_and_typecheck(
+        arena,
+        file.clone(),
+        None,
+        RocCacheDir::Persistent(cache::roc_cache_dir().as_path()),
+        load_config,
+    )?;
+
+    let buf = annotate_module(&mut loaded);
+
+    std::fs::write(&file, buf.as_str())
+        .unwrap_or_else(|e| internal_error!("failed to write annotated file to {file:?}: {e}"));
+
+    Ok(())
+}
+
+fn annotate_module(loaded: &mut LoadedModule) -> String {
+    let (decls, subs, abilities) =
+        if let Some(decls) = loaded.declarations_by_id.get(&loaded.module_id) {
+            let subs = loaded.solved.inner_mut();
+            let abilities = &loaded.abilities_store;
+
+            (decls, subs, abilities)
+        } else if let Some(checked) = loaded.typechecked.get_mut(&loaded.module_id) {
+            let decls = &checked.decls;
+            let subs = checked.solved_subs.inner_mut();
+            let abilities = &checked.abilities_store;
+
+            (decls, subs, abilities)
+        } else {
+            internal_error!("Could not find file's module");
+        };
+
+    let src = &loaded
+        .sources
+        .get(&loaded.module_id)
+        .unwrap_or_else(|| internal_error!("Could not find the file's source"))
+        .1;
+
+    let edits = annotation_edits(
+        decls,
+        subs,
+        abilities,
+        src,
+        loaded.module_id,
+        &loaded.interns,
+    );
+
+    let mut buffer = String::new();
+    let mut file_progress = 0;
+
+    for (position, edit) in edits {
+        if file_progress > position {
+            internal_error!("Module definitions are out of order");
+        }
+
+        buffer.push_str(&src[file_progress..position]);
+        buffer.push_str(&edit);
+
+        file_progress = position;
+    }
+    buffer.push_str(&src[file_progress..]);
+
+    buffer
+}
+
+pub fn annotation_edits(
```

Comment:
I've exposed this and annotations_edit() publicly to be used in the language server but It feels weird to expose them from the cli crate. If it should be in another crate what one?